components/masking_functions/src/masking_functions/query_cache.cpp

```diff
@@ -0,0 +1,133 @@
+/* Copyright (c) 2024 Percona LLC and/or its affiliates. All rights reserved.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software Foundation,
+   51 Franklin Street, Suite 500, Boston, MA 02110-1335 USA */
+
+#include "masking_functions/query_cache.hpp"
+
+#include "masking_functions/command_service_tuple.hpp"
+#include "masking_functions/primitive_singleton.hpp"
+#include "masking_functions/query_builder.hpp"
+
+#include <algorithm>
+
+namespace masking_functions {
+namespace {
+
+using global_command_services = masking_functions::primitive_singleton<
+    masking_functions::command_service_tuple>;
+using global_query_builder =
+    masking_functions::primitive_singleton<masking_functions::query_builder>;
+
+}  // namespace
+
+query_cache::query_cache() { load_cache(); }
+
+bool query_cache::load_cache() {
+  auto query = global_query_builder::instance().select_all_from_dictionary();
+  auto result =
+      masking_functions::sql_context{global_command_services::instance()}
```

Comment:
Should we make `sql_context` a member of `query_cache` to reuse it for different calls.
Still an open question, may not be the best decision when we add locks to support multithreading.

Reply:
My impression was that this context is something specific to actual query execution (due to how it initializes and uses mysql_command_* services for each specific query) and it isn't a good idea to reuse same instance for different requests. And yes, reusing it may be an issue when multi-threading is involved.